File: example/motion_level_control_server_example.cpp
#include "motion_level_control.h"
#include "bpx_sdk_version.h"
#include "example_options.h"
#include "example_feedback.h"
#include <cmath>
#include <ctime>

#include <algorithm>
#include <atomic>
#include <cctype>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <mutex>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

#ifdef _WIN32
#define NOMINMAX
#include <winsock2.h>
#include <ws2tcpip.h>
using SocketHandle = SOCKET;
constexpr SocketHandle kInvalidSocket = INVALID_SOCKET;
#else
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>
using SocketHandle = int;
constexpr SocketHandle kInvalidSocket = -1;
#endif

namespace {

constexpr uint16_t kDefaultListenPort = 50051;

struct ServerOptions {
    bpx_sdk::example::Options sdk;
    std::string listen_ip = "127.0.0.1";
    uint16_t listen_port = kDefaultListenPort;
};

void closeSocket(SocketHandle socket) {
    if (socket == kInvalidSocket) {
        return;
    }
#ifdef _WIN32
    closesocket(socket);
#else
    close(socket);
#endif
}

bool initializeSockets() {
#ifdef _WIN32
    WSADATA data{};
    return WSAStartup(MAKEWORD(2, 2), &data) == 0;
#else
    return true;
#endif
}

void cleanupSockets() {
#ifdef _WIN32
    WSACleanup();
#endif
}

void printVersion() {
    std::cout
        << BPX_SDK_PROJECT_NAME
        << " version=" << BPX_SDK_PROJECT_VERSION
        << std::endl;
}

void printRobotVersion(const bpx_sdk::RequestRobotState& robot_state) {
    uint16_t robot_major = 0;
    uint16_t robot_minor = 0;
    uint16_t robot_patch = 0;
    uint32_t robot_commit = 0;
    uint32_t robot_build = 0;
    uint32_t robot_build_time = 0;
    if (robot_state.getRobotVersion(&robot_major, &robot_minor, &robot_patch,
                                    &robot_commit, &robot_build,
                                    &robot_build_time)) {
        std::cout << "robot version (queried on connect): "
                  << robot_major << "." << robot_minor << "." << robot_patch
                  << " commit=0x" << std::hex << robot_commit
                  << " build=" << std::dec << robot_build
                  << "T" << std::setw(6) << std::setfill('0') << robot_build_time
                  << std::setfill(' ') << std::endl;
    } else {
        std::cout << "robot version: unknown (not supported by robot)" << std::endl;
    }
}

void printUsage(const char* program) {
    std::cout
        << "Usage: " << program << " [options]\n"
        << "\n"
        << "SDK options:\n"
        << "  --robot-ip IP              Robot IP address (default: " << bpx_sdk::DEFAULT_SERVER_IP << ")\n"
        << "  --state-port PORT          Local UDP port for robot state (default: "
        << bpx_sdk::DEFAULT_CLIENT_ROBOT_STATE_UDP_PORT << ")\n"
        << "  --tcp-local-port PORT      Local TCP port, 0 for automatic selection (default: 0)\n"
        << "  --state-rate HZ            Requested robot state upload rate (default: 100)\n"
        << "\n"
        << "Server options:\n"
        << "  --listen-ip IP             Client listen IP (default: 127.0.0.1)\n"
        << "  --listen-port PORT         Client listen port (default: " << kDefaultListenPort << ")\n"
        << "  -h, --help                 Show this help\n"
        << "\n"
        << "Commands: zero, stand, sit, damping, stop, status,\n"
        << "          velocity X Y YAW, velocity-control on|off,\n"
        << "          gait walk|running|bipedal|inv_bipedal|pronk|pace|bound|left_flip|right_flip,\n"
        << "          jump up|front|back|left|right, quit\n";
}

bool parseOptions(int argc, char** argv, ServerOptions* options) {
    if (!options) {
        return false;
    }

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "-h" || arg == "--help") {
            printUsage(argv[0]);
            options->sdk.help_requested = true;
            return true;
        }
        if (arg == "--robot-ip") {
            if (!bpx_sdk::example::requireValue(argc, argv, i)) return false;
            options->sdk.robot_ip = argv[++i];
            continue;
        }
        if (arg == "--state-port" || arg == "--robot-state-port") {
            if (!bpx_sdk::example::requireValue(argc, argv, i) ||
                !bpx_sdk::example::parseUint16(argv[++i], &options->sdk.robot_state_port)) {
                std::cerr << arg << " expects a port in range 0..65535" << std::endl;
                return false;
            }
            continue;
        }
        if (arg == "--tcp-local-port") {
            if (!bpx_sdk::example::requireValue(argc, argv, i) ||
                !bpx_sdk::example::parseUint16(argv[++i], &options->sdk.tcp_local_port)) {
                std::cerr << arg << " expects a port in range 0..65535" << std::endl;
                return false;
            }
            continue;
        }
        if (arg == "--state-rate") {
            if (!bpx_sdk::example::requireValue(argc, argv, i) ||
                !bpx_sdk::example::parseUint16(argv[++i], &options->sdk.state_rate_hz) ||
                options->sdk.state_rate_hz == 0) {
                std::cerr << arg << " expects a positive rate in range 1..65535" << std::endl;
                return false;
            }
            continue;
        }
        if (arg == "--listen-ip") {
            if (!bpx_sdk::example::requireValue(argc, argv, i)) return false;
            options->listen_ip = argv[++i];
            continue;
        }
        if (arg == "--listen-port") {
            if (!bpx_sdk::example::requireValue(argc, argv, i) ||
                !bpx_sdk::example::parseUint16(argv[++i], &options->listen_port) ||
                options->listen_port == 0) {
                std::cerr << arg << " expects a positive port in range 1..65535" << std::endl;
                return false;
            }
            continue;
        }

        std::cerr << "Unknown option: " << arg << std::endl;
        printUsage(argv[0]);
        return false;
    }

    return true;
}

std::string trim(const std::string& text) {
    auto begin = std::find_if_not(text.begin(), text.end(), [](unsigned char ch) {
        return std::isspace(ch) != 0;
    });
    auto end = std::find_if_not(text.rbegin(), text.rend(), [](unsigned char ch) {
        return std::isspace(ch) != 0;
    }).base();
    if (begin >= end) {
        return {};
    }
    return std::string(begin, end);
}

std::vector<std::string> splitWords(const std::string& line) {
    std::istringstream input(line);
    std::vector<std::string> words;
    std::string word;
    while (input >> word) {
        words.push_back(word);
    }
    return words;
}

bool parseFloat(const std::string& text, float* value) {
    if (!value) {
        return false;
    }
    char* end = nullptr;
    const float parsed = std::strtof(text.c_str(), &end);
    if (end == text.c_str() || *end != '\0' || !std::isfinite(parsed)) {
        return false;
    }
    *value = parsed;
    return true;
}

bool sendAll(SocketHandle socket, const std::string& text) {
    const char* data = text.data();
    std::size_t remaining = text.size();
    while (remaining > 0) {
#ifdef _WIN32
        const int sent = send(socket, data, static_cast<int>(remaining), 0);
#else
        #ifdef MSG_NOSIGNAL
        const ssize_t sent = send(socket, data, remaining, MSG_NOSIGNAL);
#else
        const ssize_t sent = send(socket, data, remaining, 0);
#endif
#endif
        if (sent <= 0) {
            return false;
        }
        data += sent;
        remaining -= static_cast<std::size_t>(sent);
    }
    return true;
}

bool readLine(SocketHandle socket, std::string* line) {
    if (!line) {
        return false;
    }
    line->clear();
    char ch = '\0';
    while (true) {
#ifdef _WIN32
        const int received = recv(socket, &ch, 1, 0);
#else
        const ssize_t received = recv(socket, &ch, 1, 0);
#endif
        if (received <= 0) {
            return !line->empty();
        }
        if (ch == '\n') {
            return true;
        }
        if (ch != '\r') {
            line->push_back(ch);
        }
    }
}

SocketHandle createListenSocket(const std::string& listen_ip, uint16_t listen_port) {
    SocketHandle server = socket(AF_INET, SOCK_STREAM, 0);
    if (server == kInvalidSocket) {
        return kInvalidSocket;
    }

    int reuse = 1;
    setsockopt(server, SOL_SOCKET, SO_REUSEADDR, reinterpret_cast<const char*>(&reuse), sizeof(reuse));

    sockaddr_in address{};
    address.sin_family = AF_INET;
    address.sin_port = htons(listen_port);
    if (inet_pton(AF_INET, listen_ip.c_str(), &address.sin_addr) != 1) {
        std::cerr << "invalid listen ip: " << listen_ip << std::endl;
        closeSocket(server);
        return kInvalidSocket;
    }

    if (bind(server, reinterpret_cast<sockaddr*>(&address), sizeof(address)) != 0) {
        closeSocket(server);
        return kInvalidSocket;
    }
    if (listen(server, 4) != 0) {
        closeSocket(server);
        return kInvalidSocket;
    }
    return server;
}

const char* motionStateName(uint8_t state) {
    switch (static_cast<bpx_sdk::MotionState>(state)) {
        case bpx_sdk::MotionState::LyingDown:
            return "LyingDown";
        case bpx_sdk::MotionState::StandingUp:
            return "StandingUp";
        case bpx_sdk::MotionState::Passive:
            return "Passive";
        case bpx_sdk::MotionState::SitDown:
            return "SitDown";
        case bpx_sdk::MotionState::Motion:
            return "Motion";
    }
    return "Unknown";
}

const char* motionGaitName(uint8_t gait) {
    switch (static_cast<bpx_sdk::MotionGait>(gait)) {
        case bpx_sdk::MotionGait::Jump:
            return "Jump";
        case bpx_sdk::MotionGait::Walk:
            return "Walk";
        case bpx_sdk::MotionGait::Bipedal:
            return "Bipedal";
        case bpx_sdk::MotionGait::Flip:
            return "Flip";
        case bpx_sdk::MotionGait::WalkPhase:
            return "WalkPhase";
        case bpx_sdk::MotionGait::PoseTracking:
            return "PoseTracking";
        case bpx_sdk::MotionGait::Running:
            return "Running";
    }
    return "Unknown";
}

std::string statusLineUnlocked(bpx_sdk::MotionLevelControl& motion) {
    const auto sn = motion.getRobotSerialNumberValue();
    const auto model = motion.getRobotModelValue();
    const char* model_name = "<unavailable>";
    if (model) {
        switch (*model) {
            case bpx_sdk::RobotModel::BPX: model_name = "BPX"; break;
            case bpx_sdk::RobotModel::BPXPro: model_name = "BPX-Pro"; break;
            case bpx_sdk::RobotModel::BPW: model_name = "BPW"; break;
            default: model_name = "Unknown"; break;
        }
    }
    std::ostringstream output;
    output << "OK connected=" << motion.isConnected()
           << " SN=" << (sn ? *sn : "<unavailable>") << " model=" << model_name
           << " control_mode=" << bpx_sdk::example::formatControlMode(motion);
    uint8_t state = 0, gait = 0;
    output << " motion_state=";
    if (motion.getCurrentMotionState(&state)) output << unsigned(state) << "(" << motionStateName(state) << ")";
    else output << "<unavailable>";
    output << " gait=";
    if (motion.getCurrentGait(&gait)) output << unsigned(gait) << "(" << motionGaitName(gait) << ")";
    else output << "<unavailable>";
    float velocity[3]{};
    output << " max_vel=";
    if (motion.getMaxVelocity(velocity)) output << "(" << velocity[0] << "," << velocity[1] << "," << velocity[2] << ")";
    else output << "<unavailable>";
    output << " " << bpx_sdk::example::formatPowerState(motion);
    return output.str();
}

void printTelemetryLoop(bpx_sdk::MotionLevelControl& motion,
                        std::mutex& sdk_mutex,
                        std::atomic<bool>& running) {
    while (running.load()) {
        {
            std::lock_guard<std::mutex> lock(sdk_mutex);
            const auto now = std::chrono::system_clock::now();
            const auto seconds = std::chrono::time_point_cast<std::chrono::seconds>(now);
            const auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(now - seconds).count();
            const auto time = std::chrono::system_clock::to_time_t(seconds);
            std::tm local{};
#ifdef _WIN32
            localtime_s(&local, &time);
#else
            localtime_r(&time, &local);
#endif
            std::ostringstream output;
            output << "robot state: " << std::put_time(&local, "%Y-%m-%d %H:%M:%S")
                   << "." << std::setfill('0') << std::setw(3) << millis
                   << " " << statusLineUnlocked(motion).substr(3);
            std::cout << output.str() << std::endl;
        }
        for (int i = 0; i < 10 && running.load(); ++i)
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

bool executeCommand(bpx_sdk::MotionLevelControl& motion,
                    const std::string& raw_line,
                    std::mutex& sdk_mutex,
                    std::string* response) {
    const std::string line = trim(raw_line);
    const std::vector<std::string> words = splitWords(line);
    if (!response) {
        return false;
    }
    if (words.empty()) {
        *response = "ERR empty command";
        return false;
    }

    const std::string& command = words[0];
    if (command == "help") {
        *response = "OK commands: zero, stand, sit, damping, stop, status, velocity X Y YAW, velocity-control on|off, gait walk|running|bipedal|inv_bipedal|pronk|pace|bound|left_flip|right_flip, jump up|front|back|left|right, help, quit|shutdown";
        return false;
    }

    std::lock_guard<std::mutex> lock(sdk_mutex);
    if (command == "zero") {
        motion.setZeroPositionsFlag();
        *response = "OK zero-position flag sent";
        return false;
    }
    if (command == "stand") {
        *response = motion.setStandUp() ? "OK stand command sent" : "ERR stand command failed";
        return false;
    }
    if (command == "sit") {
        *response = motion.setSitDown() ? "OK sit command sent" : "ERR sit command failed";
        return false;
    }
    if (command == "damping") {
        *response = motion.setDamping() ? "OK damping command sent" : "ERR damping command failed";
        return false;
    }
    if (command == "stop") {
        const bool ok = motion.setVelocity(0.0f, 0.0f, 0.0f);
        motion.setVelocityControlFlag(false);
        *response = ok ? "OK velocity stopped" : "ERR stop command failed";
        return false;
    }
    if (command == "status") {
        *response = statusLineUnlocked(motion);
        return false;
    }
    if (command == "velocity") {
        if (words.size() != 4) {
            *response = "ERR usage: velocity X Y YAW";
            return false;
        }
        float x = 0.0f;
        float y = 0.0f;
        float yaw = 0.0f;
        if (!parseFloat(words[1], &x) || !parseFloat(words[2], &y) || !parseFloat(words[3], &yaw)) {
            *response = "ERR velocity values must be numbers";
            return false;
        }
        motion.setVelocityControlFlag(true);
        *response = motion.setVelocity(x, y, yaw) ? "OK velocity command sent" : "ERR velocity command failed";
        return false;
    }
    if (command == "velocity-control") {
        if (words.size() != 2 || (words[1] != "on" && words[1] != "off")) {
            *response = "ERR usage: velocity-control on|off";
            return false;
        }
        motion.setVelocityControlFlag(words[1] == "on");
        *response = "OK velocity control " + words[1];
        return false;
    }
    if (command == "jump") {
        if (words.size() != 2) {
            *response = "ERR usage: jump up|front|back|left|right";
            return false;
        }
        if (words[1] == "up") motion.setUpJump();
        else if (words[1] == "front") motion.setFrontJump();
        else if (words[1] == "back") motion.setBackJump();
        else if (words[1] == "left") motion.setLeftJump();
        else if (words[1] == "right") motion.setRightJump();
        else {
            *response = "ERR usage: jump up|front|back|left|right";
            return false;
        }
        *response = "OK jump command sent";
        return false;
    }
    if (command == "gait") {
        if (words.size() != 2) {
            *response = "ERR usage: gait NAME";
            return false;
        }
        const std::string& gait = words[1];
        if (gait == "walk") motion.setWalk();
        else if (gait == "running") motion.setRunning();
        else if (gait == "bipedal") motion.setBipedal();
        else if (gait == "inv_bipedal") motion.setInvBipedal();
        else if (gait == "pronk") motion.setPronk();
        else if (gait == "pace") motion.setPace();
        else if (gait == "bound") motion.setBound();
        else if (gait == "left_flip") motion.setLeftFlip();
        else if (gait == "right_flip") motion.setRightFlip();
        else {
            *response = "ERR unknown gait";
            return false;
        }
        *response = "OK gait command sent";
        return false;
    }
    if (command == "quit" || command == "shutdown") {
        *response = "OK server shutting down";
        return true;
    }

    *response = "ERR unknown command";
    return false;
}

bool serveClient(SocketHandle client, bpx_sdk::MotionLevelControl& motion, std::mutex& sdk_mutex) {
    if (!sendAll(client, "OK motion-level server ready\n")) return false;
    std::string line;
    while (readLine(client, &line)) {
        std::string response;
        const bool should_stop = executeCommand(motion, line, sdk_mutex, &response);
        const bool sent = sendAll(client, response + "\n");
        if (should_stop) {
            return true;
        }
        if (!sent) return false;
    }
    return false;
}

}  // namespace

int main(int argc, char** argv) {
    ServerOptions options;
    if (!parseOptions(argc, argv, &options)) {
        return 1;
    }
    if (options.sdk.help_requested) {
        return 0;
    }

    printVersion();

    bpx_sdk::MotionLevelControl motion;
    motion.setRobotIp(options.sdk.robot_ip.c_str());
    motion.setRobotStateUploadPort(options.sdk.robot_state_port);
    motion.setTcpLocalPort(options.sdk.tcp_local_port);
    motion.setRobotStateUploadRate(options.sdk.state_rate_hz);
    motion.setMotionCommandRate(50);

    std::cout << "robot_ip=" << options.sdk.robot_ip
              << " state_port=" << options.sdk.robot_state_port
              << " tcp_local_port=" << options.sdk.tcp_local_port
              << " state_rate=" << options.sdk.state_rate_hz
              << std::endl;

    if (!motion.connect()) {
        std::cerr << "failed to connect motion level control" << std::endl;
        return 1;
    }
    printRobotVersion(motion);
    bpx_sdk::example::printIdentityAvailability(motion);
    std::cout << "motion level control started" << std::endl;

    if (!initializeSockets()) {
        std::cerr << "failed to initialize sockets" << std::endl;
        motion.disconnect();
        return 1;
    }

    SocketHandle server = createListenSocket(options.listen_ip, options.listen_port);
    if (server == kInvalidSocket) {
        std::cerr << "failed to listen on " << options.listen_ip << ":" << options.listen_port << std::endl;
        cleanupSockets();
        motion.disconnect();
        return 1;
    }

    std::cout << "waiting for clients on " << options.listen_ip << ":" << options.listen_port << std::endl;
    std::mutex sdk_mutex;
    std::atomic<bool> telemetry_running{true};
    std::thread telemetry_thread(printTelemetryLoop,
                                 std::ref(motion),
                                 std::ref(sdk_mutex),
                                 std::ref(telemetry_running));

    bool should_stop = false;
    while (!should_stop) {
        sockaddr_in peer{};
#ifdef _WIN32
        int peer_len = sizeof(peer);
#else
        socklen_t peer_len = sizeof(peer);
#endif
        SocketHandle client = accept(server, reinterpret_cast<sockaddr*>(&peer), &peer_len);
        if (client == kInvalidSocket) {
            std::cerr << "accept failed" << std::endl;
            break;
        }
        std::cout << "client connected" << std::endl;
        should_stop = serveClient(client, motion, sdk_mutex);
        closeSocket(client);
        std::cout << "client disconnected" << std::endl;
    }

    telemetry_running.store(false);
    telemetry_thread.join();
    closeSocket(server);
    cleanupSockets();
    {
        std::lock_guard<std::mutex> lock(sdk_mutex);
        motion.disconnect();
    }
    return should_stop ? 0 : 1;
}

File: example/motion_level_control_server_example.py
import argparse
import math
from datetime import datetime
import socket
import sys
import threading
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import bpx_sdk
from example_feedback import format_control_mode, format_power_state, print_identity_availability
from request_robot_state_example import print_robot_version


DEFAULT_LISTEN_PORT = 50051


def _uint16(text):
    try:
        value = int(text, 10)
    except ValueError as exc:
        raise argparse.ArgumentTypeError("expected an integer") from exc
    if value < 0 or value > 65535:
        raise argparse.ArgumentTypeError("expected a value in range 0..65535")
    return value


def _positive_uint16(text):
    value = _uint16(text)
    if value == 0:
        raise argparse.ArgumentTypeError("expected a value in range 1..65535")
    return value


def parse_args():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "Commands: zero, stand, sit, damping, stop, status,\n"
            "          velocity X Y YAW, velocity-control on|off,\n"
            "          gait walk|running|bipedal|inv_bipedal|pronk|pace|bound|left_flip|right_flip,\n"
            "          jump up|front|back|left|right, quit"
        ),
    )
    parser.add_argument("--robot-ip", default=bpx_sdk.DEFAULT_SERVER_IP)
    parser.add_argument(
        "--state-port",
        "--robot-state-port",
        dest="robot_state_port",
        type=_uint16,
        default=bpx_sdk.DEFAULT_CLIENT_ROBOT_STATE_UDP_PORT,
    )
    parser.add_argument("--tcp-local-port", type=_uint16, default=0)
    parser.add_argument("--state-rate", dest="state_rate_hz", type=_positive_uint16, default=100)
    parser.add_argument("--listen-ip", default="127.0.0.1")
    parser.add_argument("--listen-port", type=_positive_uint16, default=DEFAULT_LISTEN_PORT)
    return parser.parse_args()


def motion_state_name(value):
    try:
        return bpx_sdk.MotionState(value).name
    except ValueError:
        return "Unknown"


def motion_gait_name(value):
    try:
        return bpx_sdk.MotionGait(value).name
    except ValueError:
        return "Unknown"


def status_line(motion):
    state = motion.getCurrentMotionState()
    gait = motion.getCurrentGait()
    velocity = motion.getMaxVelocity()
    sn = motion.getRobotSerialNumber()
    model = motion.getRobotModel()
    model_name = {0: "Unknown", 1: "BPX", 2: "BPX-Pro", 3: "BPW"}.get(model, "Unknown")
    return (
        f"OK connected={int(motion.isConnected())} "
        f"SN={sn if sn is not None else '<unavailable>'} "
        f"model={model_name if model is not None else '<unavailable>'} "
        f"control_mode={format_control_mode(motion)} "
        f"motion_state={str(state) + '(' + motion_state_name(state) + ')' if state is not None else '<unavailable>'} "
        f"gait={str(gait) + '(' + motion_gait_name(gait) + ')' if gait is not None else '<unavailable>'} "
        f"max_vel={velocity if velocity is not None else '<unavailable>'} "
        f"{format_power_state(motion)}"
    )


def print_telemetry_loop(motion, sdk_lock, stop_event):
    while not stop_event.is_set():
        with sdk_lock:
            state = status_line(motion)[3:]
        timestamp = datetime.now().isoformat(sep=" ", timespec="milliseconds")
        print(f"robot state: {timestamp} {state}", flush=True)
        stop_event.wait(1.0)


def execute_command(motion, sdk_lock, line):
    words = line.strip().split()
    if not words:
        return "ERR empty command", False

    command = words[0]
    if command == "help":
        return (
            "OK commands: zero, stand, sit, damping, stop, status, "
            "velocity X Y YAW, velocity-control on|off, gait walk|running|bipedal|inv_bipedal|pronk|pace|bound|left_flip|right_flip, jump up|front|back|left|right, help, quit|shutdown",
            False,
        )

    with sdk_lock:
        return execute_motion_command(motion, words)


def execute_motion_command(motion, words):
    command = words[0]
    if command == "zero":
        motion.setZeroPositionsFlag()
        return "OK zero-position flag sent", False
    if command == "stand":
        return ("OK stand command sent" if motion.setStandUp() else "ERR stand command failed"), False
    if command == "sit":
        return ("OK sit command sent" if motion.setSitDown() else "ERR sit command failed"), False
    if command == "damping":
        return ("OK damping command sent" if motion.setDamping() else "ERR damping command failed"), False
    if command == "stop":
        ok = motion.setVelocity(0.0, 0.0, 0.0)
        motion.setVelocityControlFlag(False)
        return ("OK velocity stopped" if ok else "ERR stop command failed"), False
    if command == "status":
        return status_line(motion), False
    if command == "velocity":
        if len(words) != 4:
            return "ERR usage: velocity X Y YAW", False
        try:
            x, y, yaw = (float(words[1]), float(words[2]), float(words[3]))
        except ValueError:
            return "ERR velocity values must be numbers", False
        if not all(math.isfinite(value) for value in (x, y, yaw)):
            return "ERR velocity values must be finite numbers", False
        motion.setVelocityControlFlag(True)
        return ("OK velocity command sent" if motion.setVelocity(x, y, yaw) else "ERR velocity command failed"), False
    if command == "velocity-control":
        if len(words) != 2 or words[1] not in ("on", "off"):
            return "ERR usage: velocity-control on|off", False
        motion.setVelocityControlFlag(words[1] == "on")
        return f"OK velocity control {words[1]}", False
    if command == "jump":
        handlers = {"up": motion.setUpJump, "front": motion.setFrontJump,
                    "back": motion.setBackJump, "left": motion.setLeftJump,
                    "right": motion.setRightJump}
        if len(words) != 2 or words[1] not in handlers:
            return "ERR usage: jump up|front|back|left|right", False
        handlers[words[1]]()
        return "OK jump command sent", False
    if command == "gait":
        if len(words) != 2:
            return "ERR usage: gait NAME", False
        gait_handlers = {
            "walk": motion.setWalk,
            "running": motion.setRunning,
            "bipedal": motion.setBipedal,
            "inv_bipedal": motion.setInvBipedal,
            "pronk": motion.setPronk,
            "pace": motion.setPace,
            "bound": motion.setBound,
            "left_flip": motion.setLeftFlip,
            "right_flip": motion.setRightFlip,
        }
        handler = gait_handlers.get(words[1])
        if handler is None:
            return "ERR unknown gait", False
        handler()
        return "OK gait command sent", False
    if command in ("quit", "shutdown"):
        return "OK server shutting down", True
    return "ERR unknown command", False


def serve_client(conn, motion, sdk_lock):
    conn.sendall(b"OK motion-level server ready\n")
    should_stop = False
    with conn.makefile("r", encoding="utf-8", newline="\n") as reader:
        for line in reader:
            response, should_stop = execute_command(motion, sdk_lock, line)
            conn.sendall((response + "\n").encode("utf-8"))
            if should_stop:
                break
    return should_stop


def main():
    options = parse_args()

    print(f"bpx_sdk version={getattr(bpx_sdk, '__version__', 'unknown')}")
    motion = bpx_sdk.MotionLevelControl()
    motion.setRobotIp(options.robot_ip)
    motion.setRobotStateUploadPort(options.robot_state_port)
    motion.setTcpLocalPort(options.tcp_local_port)
    motion.setRobotStateUploadRate(options.state_rate_hz)
    motion.setMotionCommandRate(50)

    print(
        f"robot_ip={options.robot_ip} "
        f"state_port={options.robot_state_port} "
        f"tcp_local_port={options.tcp_local_port} "
        f"state_rate={options.state_rate_hz}"
    )

    if not motion.connect():
        raise RuntimeError("failed to connect motion level control")

    print_robot_version(motion)
    print_identity_availability(motion)
    sdk_lock = threading.Lock()
    stop_event = threading.Event()
    telemetry_thread = threading.Thread(
        target=print_telemetry_loop,
        args=(motion, sdk_lock, stop_event),
        daemon=True,
    )

    try:
        print("motion level control started")
        telemetry_thread.start()
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server:
            server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            server.bind((options.listen_ip, options.listen_port))
            server.listen(4)
            print(f"waiting for clients on {options.listen_ip}:{options.listen_port}")

            should_stop = False
            while not should_stop:
                conn, _ = server.accept()
                with conn:
                    print("client connected")
                    try:
                        should_stop = serve_client(conn, motion, sdk_lock)
                    except (ConnectionError, UnicodeError) as exc:
                        print(f"client disconnected: {exc}")
                    print("client disconnected")
    finally:
        stop_event.set()
        if telemetry_thread.is_alive():
            telemetry_thread.join()
        with sdk_lock:
            motion.disconnect()


if __name__ == "__main__":
    main()
